keps/sig-api-machinery/20200123-client-go-ctx.md

```diff
@@ -0,0 +1,188 @@
+		Resource("pods").
+		Name(name).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Do().
+		Into(result)
+	return
+}
+```
+
+Here we extend the signature to add `context.Context` as the first argument,
+then attach the context to the request.
+
+Initially `lister` and `informer` APIs (both generated and generic) will not
+change.
+
+### Risks and Mitigations
+
+This is a breaking change that would likely require extensive changes to both
+tests and implementation code throughout the impacted code bases when client-go
+is upgraded.
+
+Inside the Kubernetes codebase, we will refactor all callsites to pass
+`context.TODO()` in the initial pass.
+
+TODO(mikedanese): Discuss and document migration path for external codebases
+that depend on client-go.
+
+### Graduation Criteria
+
+This enhancement will be GA immediately.
+
+## Alternatives
+
+The client resources (Pods, Secrets etc...) each expose a resource specific
+interface `PodInterface` for example. This is returned from a `Getter`
+interface:
+
+```go
+type PodsGetter interface {
+	Pods(namespace string) PodInterface
+}
+``` 
+
+We can modify these resource interfaces and their underlying concrete types,
+`pod` in this case, to add a `WithContext` method. This change would be
+backwards compatible for most consumers of the API although it would break any
+implementations of `PodsGetter` (e.g. in unit tests) outside of
+k8s.io/client-go.
+
+```go
+// PodInterface has methods to work with Pod resources.
+type PodInterface interface {
+  	WithContext(ctx context.Context) PodInterface
+    ...
+}
+
+type pods struct {
+	client rest.Interface
+	ns     string
+	ctx    context.Context
+}
+
+// WithContext allows you to set a context that will be used by the underlying http request
+func (c *pods) WithContext(ctx context.Context) PodInterface {
+  	c.ctx = ctx
+  	return c
+}
+
+```
+
+To pass through this context, it would be necessary to change the underlying
+client calls to accept the context. Example:
+
+```go 
+func (c *pods) Get(name string, options meta_v1.GetOptions) (result *v1.Pod, err error) {
+	result = &v1.Pod{}
+	err = c.client.Get().
+		Context(c.ctx).
+		Namespace(c.ns).
+		Resource("pods").
+		Name(name).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Do().
+		Into(result)
+	return
+}
+```
+
+This would result in an API that would be interacted with like so if the
+consumer wished to pass a context:
+
+```go  
+   ctx := req.Context()
+   pod, err := k8client.CoreV1().Pods(namespace).WithContext(ctx).Get(podName, ...)
+```
+
+This approach was rejected because:
+
+- The idea behind using a context with a request is it is only meant to live as
+  long as the initial request that created it. It is not meant to be stored. If we
+  expose an error-prone API like the one outlined, the consumer would need to call
+  `WithContext` on each request if a previous caller had set a context. This API
+  allows broken code to compile without actually improve the migration story.
```

Comment:
```suggestion
  allows broken code to compile but would fail at runtime if multiple requests were made with the same client whose context had expired.
```